pkg/visitors/tightenme.go

```diff
@@ -0,0 +1,121 @@
+func (v *TightenME) Visit(f uefi.Firmware) error {
+	switch f := f.(type) {
+	case *uefi.FlashDescriptor:
+		v.fd = f
+		return nil
+	case *uefi.MERegion:
+		v.mer = f
+		return nil
+	case *uefi.BIOSRegion:
+		v.br = f
+		return nil
+	default:
+		return f.ApplyChildren(v)
+	}
+}
+
+func (v *TightenME) process() error {
+	// Ensuring IFD exists also ensure FRegion are populated in Regions
+	if v.fd == nil {
+		return fmt.Errorf("no IFD found")
+	}
+	if v.mer == nil {
+		return fmt.Errorf("no ME region found")
+	}
+	if v.br == nil {
+		return fmt.Errorf("no BIOS region found")
+	}
+	if v.mer.FRegion.EndOffset() != v.br.FRegion.BaseOffset() {
+		return fmt.Errorf("ME and BIOS regions are not contiguous: ME end at %#x BIOS starts at %#x", v.mer.FRegion.EndOffset(), v.br.FRegion.BaseOffset())
+
+	}
+	// Compute the new limit
+	updateOffset := uint64(v.mer.FRegion.BaseOffset()) + v.mer.FreeSpaceOffset
+	updateBase := (updateOffset + uefi.RegionBlockSize - 1) / uefi.RegionBlockSize
+	updateOffset = updateBase * uefi.RegionBlockSize
+	bufOffset := updateOffset - uint64(v.mer.FRegion.BaseOffset())
+	// check the zone if empty
+	buf := v.mer.Buf()
+	if !uefi.IsErased(buf[bufOffset:], uefi.Attributes.ErasePolarity) {
+		return fmt.Errorf("ME unused space in not erased as expected")
+	}
+	// Shrink ME Region
+	v.mer.FRegion.Limit = uint16(updateBase - 1)
+	v.mer.SetBuf(buf[:bufOffset])
+	// Expand BIOS Region
+	offsetShift := uint64(v.br.FRegion.BaseOffset()) - updateOffset
+	fmt.Printf("JVDG %#x %#x %#x\n", v.br.FRegion.BaseOffset(), updateOffset, offsetShift)
+	v.br.FRegion.Base = uint16(updateBase)
+	v.br.Length += offsetShift
+	//  update elements offsets
+	for i, e := range v.br.Elements {
+		switch f := e.Value.(type) {
+		case *uefi.FirmwareVolume:
+			f.FVOffset += offsetShift
+		case *uefi.BIOSPadding:
+			f.Offset += offsetShift
+		default:
+			return fmt.Errorf("Unexpected Element at %d: %s", i, e.Type)
+		}
+	}
+	//  insert BIOSPad
+	bp, err := uefi.NewBIOSPadding(buf[bufOffset:], 0)
+	if err != nil {
+		return fmt.Errorf("Could not create BIOS Padding: %v", err)
+	}
+	v.br.Elements = append([]*uefi.TypedFirmware{uefi.MakeTyped(bp)}, v.br.Elements...)
+	// Update IFD
+	// I though regions had references to the IFD... but it seams not so:
+	v.fd.Region.FlashRegions[uefi.RegionTypeBIOS] = *v.br.FRegion
+	v.fd.Region.FlashRegions[uefi.RegionTypeME] = *v.mer.FRegion
+	// Assemble is not regenerating IFD so update regions here
+	regions := new(bytes.Buffer)
+	err = binary.Write(regions, binary.LittleEndian, v.fd.Region)
+	if err != nil {
+		return fmt.Errorf("unable to construct binary region of IDF: got %v", err)
+	}
+
+	buf = v.fd.Buf()
+	copy(buf[v.fd.RegionStart:v.fd.RegionStart+uint(uefi.FlashRegionSectionSize)], regions.Bytes())
+	v.fd.SetBuf(buf)
+
+	return nil
+}
+
+func init() {
+	RegisterCLI("tighten-me", "tighten ME's belt to give more room for LinuxBoot", 0, func(args []string) (uefi.Visitor, error) {
```

Comment:
I love the name, but use an underscore `tighten_me` to be consistent with other visitors.

Reply:
Done.
Cool ! I wanted a name that would not induce a process like "me cleaner" that removes stuff from ME.

Here If you binary edit the IFD to change the Limit for ME and Base for BIOS it will just do that.

The rest of the code is just making sure the state on utk's internal remains valid and we can chain `tighten_me` with let's say `create-fv` and then the rest of typical linuxboot dxe editing.